File: dsa-problems/Greedy/solutions/javascript/GRD-003-festival-stall-placement.js
const readline = require("readline");

class Solution {
  maxStalls(stalls, d) {
    // Sort by end time
    stalls.sort((a, b) => a[1] - b[1]);
    
    let count = 0;
    let lastEnd = -Infinity;
    
    for (const [start, end] of stalls) {
      if (start - lastEnd >= d) {
        count++;
        lastEnd = end;
      }
    }
    
    return count;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(line.trim()));
rl.on("close", () => {
  if (data.length === 0) return;
  
  let ptr = 0;
  const [n, d] = data[ptr++].split(" ").map(Number);

  const stalls = [];
  for (let i = 0; i < n; i++) {
    const [start, end] = data[ptr++].split(" ").map(Number);
    stalls.push([start, end]);
  }

  const solution = new Solution();
  console.log(solution.maxStalls(stalls, d));
});

File: dsa-problems/Greedy/solutions/python/GRD-003-festival-stall-placement.py
import sys

def max_stalls(stalls: list, d: int) -> int:
    # Sort by end time (second element)
    stalls.sort(key=lambda x: x[1])
    
    count = 0
    last_end = -float('inf')
    
    for start, end in stalls:
        if start - last_end >= d:
            count += 1
            last_end = end
            
    return count

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
        
    iterator = iter(data)
    n = int(next(iterator))
    d = int(next(iterator))
    
    stalls = []
    for _ in range(n):
        start = int(next(iterator))
        end = int(next(iterator))
        stalls.append([start, end])

    result = max_stalls(stalls, d)
    print(result)

if __name__ == "__main__":
    main()
